test/transcoding/IntelFPGAMemoryAccesses.ll

```diff
@@ -0,0 +1,218 @@
+;   x = __builtin_intel_fpga_mem(A, 1 | 2, 0);
+;   y = __builtin_intel_fpga_mem(B, 4 | 8, 0);
+;   z = __builtin_intel_fpga_mem(C, 2, 127);
+;   x = __builtin_intel_fpga_mem(&C->Field1, 1 | 2, 127);
+;   y = __builtin_intel_fpga_mem(&C->Field2, 0, 127);
+;   z = __builtin_intel_fpga_mem(C, 1 | 2 | 4 | 8, 127);
+; }
+;
+; template <typename name, typename Func>
+; __attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
+;   kernelFunc();
+; }
+;
+; int main() {
+;   kernel_single_task<class fake_kernel>([]() {
+;     float *A;
+;     int *B;
+;     State *C;
+;     foo(A, B, C);
+;   });
+;   return 0;
+; }
+
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv %t.bc --spirv-ext=+SPV_INTEL_fpga_memory_accesses -o %t.spv
+; RUN: llvm-spirv %t.spv -to-text -o - | FileCheck %s --check-prefix=CHECK-SPIRV
+
+; RUN: llvm-spirv -r %t.spv -o %t.rev.bc
+; RUN: llvm-dis < %t.rev.bc | FileCheck %s --check-prefix=CHECK-LLVM
+
+; CHECK-SPIRV: Capability FPGAMemoryAccessesINTEL
+; CHECK-SPIRV: Extension "SPV_INTEL_fpga_memory_accesses"
+; CHECK-SPIRV: Decorate {{[0-9]+}} BurstCoalesceINTEL
+; CHECK-SPIRV: Decorate {{[0-9]+}} DontStaticallyCoalesceINTEL
+; CHECK-SPIRV: Decorate {{[0-9]+}} PrefetchINTEL 0
+; CHECK-SPIRV: Decorate {{[0-9]+}} CacheSizeINTEL 0
+; CHECK-SPIRV: Decorate {{[0-9]+}} CacheSizeINTEL 127
+; Check that the semantically meaningless decoration was
+; translated as a mere annotation
+; CHECK-SPIRV: MemberDecorate {{[0-9]+}} 1 UserSemantic "{params:0}{cache-size:127}"
+; Check regular struct member decorations
+; CHECK-SPIRV: MemberDecorate {{[0-9]+}} 0 BurstCoalesceINTEL
+; CHECK-SPIRV: MemberDecorate {{[0-9]+}} 0 CacheSizeINTEL 127
+
+target datalayout = "e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-n8:16:32:64"
+target triple = "spir64-unknown-unknown-sycldevice"
+
+%"class._ZTSZ4mainE3$_0.anon" = type { i8 }
+%struct._ZTS5State.State = type { float, i32 }
+
+; CHECK-LLVM: [[PARAM_3_CACHE_0:@\.str[0-9_.]*]] = {{.*}}{params:3}{cache-size:0}
+@.str = private unnamed_addr constant [25 x i8] c"{params:3}{cache-size:0}\00", section "llvm.metadata"
+@.str.1 = private unnamed_addr constant [14 x i8] c"<invalid loc>\00", section "llvm.metadata"
+; "params" bitmask doesn't hold 0x2 (cache size ON), so cache-size can be dropped
+; during translation
+; CHECK-LLVM: [[PARAM_12_CACHE_0:@\.str[0-9_.]*]] = {{.*}}{params:12}
+@.str.2 = private unnamed_addr constant [26 x i8] c"{params:12}{cache-size:0}\00", section "llvm.metadata"
+; CHECK-LLVM: [[PARAM_2_CACHE_127:@\.str[0-9_.]*]] = {{.*}}{params:2}{cache-size:127}
+@.str.3 = private unnamed_addr constant [27 x i8] c"{params:2}{cache-size:127}\00", section "llvm.metadata"
+; CHECK-LLVM: [[PARAM_3_CACHE_127:@\.str[0-9_.]*]] = {{.*}}{params:3}{cache-size:127}
+@.str.4 = private unnamed_addr constant [27 x i8] c"{params:3}{cache-size:127}\00", section "llvm.metadata"
+; Since "params" bitmask is set to 0, the next string isn't required to be preserved
```

Comment:
Do we need to CHECK-NOT something here then?

Reply:
As it stands, we'd better not: the obsolete annotation still gets translated forwards and backwards as `annotate` would be, without affecting SPIR-V semantics. Moreover, the FPGA BE is going to disregard the obsolete annotations, thus no stability issues are expected.

Since this is more of an implementation detail and not a requirement, I chose to go without positive or negative checks for this particular case.